### PR TITLE
[Spark] Add type widening tests for data skipping

### PR DESCRIPTION
**Pull request by johanl-db**
## Description
Add tests covering stats and data skipping when a type change is applied to a Delta table.
Covers:
- A combination of disabling/enabling storing JSON stats in checkpoint files and the impact on data skipping.
- Parsed partition values stored as part of the checkpoint file when the type of the partition column is changed.

## How was this patch tested?
See above.

### CODE REVIEW

**Comment by felipepessoto** [comment]
Do we have any tests for OptimizeMetadataOnlyDeltaQuery + Type Widening? If not, do you think we need it?
https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

**Review by fred-db — COMMENTED** [review]
Nice! :)

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 1]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+    assert(getJsonStatsType(checkpoint) === expectedJsonStatsType)
+
+    val expectedPartitionStats = if (partitioned && structStatsEnabled) Some(colType) else None
+    assert(getPartitionValuesType(checkpoint, colName) === expectedPartitionStats)
+  }
+
+  /**
+   * Reads the test table filtered by the given value and checks that files are skipped as expected.
+   */
+  def checkFileSkipping(value: Any, expectedFilesRead: Long): Unit = {
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: Right, I renamed it

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 2]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+
+    var checkpoint = getLatestCheckpoint
+    // Ensure there's no new checkpoint after the type change.
+    assert(getLatestCheckpoint.semanticEquals(initialCheckpoint))
+
+    val canSkipFiles = jsonStatsEnabled || partitioned
+
+    // The last file added isn't part of the checkpoint, it always has stats that can be used for
+    // skipping even when checkpoint stats can't be used for skipping.
+    checkFileSkipping(value = 1, if (canSkipFiles) 1 else 2)
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: 👍

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 3]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+    checkAnswer(readDeltaTable(tempPath).filter("a = 1"), Row(1, 1))
+
+    checkFileSkipping(value = Int.MinValue, if (canSkipFiles) 1 else 3)
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: 👍

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 4]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+    checkAnswer(readDeltaTable(tempPath).filter(s"a = ${Int.MinValue}"), Row(Int.MinValue, 1))
+
+    // Trigger a new checkpoint after the type change and re-check data skipping.
+    deltaLog.checkpoint()
+    checkpoint = getLatestCheckpoint
+    assert(!checkpoint.semanticEquals(initialCheckpoint))
+    checkCheckpointStats(
+      checkpoint, "a", IntegerType, partitioned, jsonStatsEnabled, structStatsEnabled)
+    // When checkpoint stats are completely disabled, the last file added can't be skipped anymore.
+    checkFileSkipping(value = 1, if (canSkipFiles) 1 else 3)
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: 👍

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 5]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+    checkFileSkipping(value = Int.MinValue, if (canSkipFiles) 1 else 3)
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: 👍

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 6]
```diff
@@ -635,14 +641,168 @@ trait DeltaTypeWideningMetadataTests {
...
+
+  for {
+    partitioned <- BOOLEAN_DOMAIN
+    jsonStatsEnabled <- BOOLEAN_DOMAIN
+    structStatsEnabled <- BOOLEAN_DOMAIN
+  }
+  testStats(s"metadata-only query, partitioned=$partitioned", jsonStatsEnabled,
+      structStatsEnabled) {
+    val partitionStr = if (partitioned) "PARTITIONED BY (a)" else ""
+    sql(s"""
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala]: True, I moved the table creation in `testStats`

**Review by fred-db — APPROVED** [review]
LGTM!

**Review by tdas — APPROVED** [review]

**Pull request merged**